src/awkward/contents/numpyarray.py

```diff
@@ -1044,6 +1044,17 @@ def _reduce_next(
         keepdims,
         behavior,
     ):
+        # HACK: switch reducer for JAX equivalent
+        # NOTE: this really needs to stay here for now, i.e. can't go in
+        # ak._do (because internal code instantiates Reducer instances)
+        # Ideally we would use the kernel mechanism for this, but JAX arrays are immutable whilst out kernel API assumes
+        # that arrays are mutable and that the inputs/outputs have a common nplike. For JAX we would probably want
+        # both numpy/cupy and jax.numpy which are not directly compatible, so this "hack" will likely remain.
+        if isinstance(self.backend, ak._backends.JaxBackend):
+            from awkward._connect.jax.reducers import get_jax_reducer
+
+            reducer = get_jax_reducer(reducer)
+
```

Comment:
JAX is very special with regard to reducers. Having it make this much noise here in the codebase doesn't bother me.

(It's only a hack in the sense that Content code should not be asking questions about specific backends, to separate those concerns, but JAX reducers stand out enough as a thing that should get special attention that I wouldn't call this a hack.[^1])

[^1]: Well, maybe in a technical sense. I've been converging on the idea that the word "to hack" means to break levels of abstraction.[^2] For instance, if you use the fact that CPython's `id` function returns C function pointers to `PyObject*` as a number, you are committing a hack because the Python language is supposed to be a different level of abstraction than its implementation. So if we declare that Content implementations are at a layer of abstraction above backend implementations, then doing anything in Content that acts differently for specific backends would technically be a hack. Content and backends/nplikes weren't originally conceived of as different layers of abstraction, but they're growing toward that direction.

[^2]: And so puns are language hacks. The meanings of words is a different level of abstraction from the way that they sound, so if a similar-sounding word is used in place of another to change the meaning of a sentence, it's mixing levels of abstraction. Some things in type theory are even called "puns" when they take advantage of unintended equivalences.

Reply:
I hope I haven't started you on a large project! We can get there incrementally.